fix: implement two-layer caching for Azure authorization and error handling

arxhive opened the pull request:
## Summary

Implements a two-layer caching architecture for Azure authorization and fixes deterministic error handling to enable caching at both Guard and Kubernetes API server levels.

## Problem

Previously, Guard only cached successful Azure CheckAccess responses. When Azure returned 4xx errors (e.g., InvalidSubscriptionId), every subsequent request made a new API call because:

1. Guard returned HTTP 5xx for Azure errors
2. Kubernetes API server only caches HTTP 200 responses
3. Neither layer cached error scenarios

## Solution

### Two-Layer Caching Architecture

```
┌─────────────────┐     ┌─────────────────┐     ┌─────────────────┐
│  Kubernetes     │     │     Guard       │     │  Azure RBAC     │
│  API Server     │     │   (BigCache)    │     │  CheckAccess    │
└────────┬────────┘     └────────┬────────┘     └────────┬────────┘
         │                       │                       │
    Layer 2 Cache          Layer 1 Cache           Azure API
    (5min TTL)             (3min TTL)
```

### Error Handling Strategy

| Azure Response | Guard Behavior | Cached? |
|----------------|----------------|---------|
| Success (allowed) | Return `{Allowed: true}` | Yes, both layers |
| Success (denied) | Return `{Allowed: false, Denied: true}` | Yes, both layers |
| 4xx Error | Return `{Allowed: false, Denied: true}` | Yes, both layers |
| 5xx Error | Return HTTP 5xx error | No (transient) |

Guard now converts deterministic 4xx errors into HTTP 200 with `Denied: true`, enabling caching at both layers.

## Changes

### Core Changes

1. **authz/providers/azure/azure.go**
   - Return cached denials for 4xx errors instead of HTTP 5xx
   - Added comprehensive package documentation

2. **authz/providers/azure/rbac/checkaccessreqhelper.go**
   - Added `CheckAccessErrorVerdict` constant

3. **authz/providers/azure/data/datastore.go**
   - Added `guard_azure_authz_cache_errors_cached_total` Prometheus metric
   - TTL reduced from 10 to 3 minutes

4. **authz/providers/azure/options/options.go**
   - Default cache TTL changed from 10 to 3 minutes

### Documentation

5. **authz/providers/azure/CACHING.md**
   - Two-layer caching architecture explanation
   - Request flow diagram
   - Error handling table
   - Monitoring metrics reference
   - Recommended configuration

### Testing Infrastructure

6. **tests/k6/** - k6 load testing framework
   - Scenarios: cache-warmup, sustained-load, burst-load, cache-eviction
   - Mock Azure server for local testing
   - TLS certificates for mTLS testing
   - Comparison scripts for performance analysis

7. **tests/mock-server/** - Mock Azure CheckAccess server
   - Simulates Azure RBAC responses
   - Configurable latency and error rates

## Configuration

```yaml
# Guard flags
--azure.cache-size-mb=50
--azure.cache-ttl-minutes=3

# Kubernetes API server flags (existing)
--authorization-webhook-cache-authorized-ttl=5m
--authorization-webhook-cache-unauthorized-ttl=5m
```

## Trade-offs & Accepted Risks

1. **3-minute TTL**: Shorter than previous 10 minutes. Reduces cache staleness but increases Azure API calls. Mitigated by API server's 5-minute cache layer.

2. **Error caching**: 4xx errors cached as denials. May mask configuration issues temporarily, but provides significant performance benefit. Transient 5xx errors are NOT cached.

3. **Two-layer dependency**: Full benefit requires Kubernetes API server webhook cache configuration. Already configured in AKS deployments.

## Metrics

New Prometheus metric for observability:
- `guard_azure_authz_cache_errors_cached_total` - Count of 4xx errors cached as denied

weinong commented on authz/providers/azure/data/datastore.go:
```diff
@@ -74,10 +94,12 @@ func (s *DataStore) Get(key string, value interface{}) (found bool, err error) {
 			return false, nil
 		}
 		stats := s.cache.Stats()
+		cacheMisses.Inc()
```
i think you just need to increment the counter at L97?

weinong commented on server/server.go:
```diff
@@ -211,8 +211,17 @@ func (s Server) ListenAndServe() {
 		m.Post("/subjectaccessreviews", authzPromHandler.ServeHTTP)
 
 		if s.AuthzRecommendedOptions.AuthzProvider.Has(azure.OrgType) {
-			options := data.DefaultOptions
```
Is `DefaultOptions` still being used? If not, remove?

arxhive commented on authz/providers/azure/data/datastore.go:
```diff
@@ -25,12 +25,32 @@ import (
 
+var (
+	cacheHits = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "guard_azure_cache_hits_total",
```
Good idea, applied!

arxhive commented:
Addressed review comments:

**Comment 2608644584** (cache miss counter): Consolidated duplicate cacheMisses.Inc() calls into a single increment at the error entry point. This reduces redundancy and ensures consistent metrics across all cache miss scenarios.

**Comment 2608695183** (DefaultOptions): Replaced the DefaultOptions variable with NewOptions() constructor function (commit af09fb1c) for better encapsulation and following Go idioms.

**Comment 2608481320** (TTL implications): The cache TTL is fully configurable via CLI flags (--azure.cache-ttl-minutes), allowing operators to tune based on their security requirements. The default value in NewOptions() serves as a fallback. The added metrics help monitor cache effectiveness to inform these decisions.

weinong commented on server/server.go:
```diff
@@ -211,8 +211,17 @@ func (s Server) ListenAndServe() {
 		if s.AuthzRecommendedOptions.AuthzProvider.Has(azure.OrgType) {
-			options := data.DefaultOptions
-			authzhandler.Store, err = data.NewDataStore(options)
+			cacheOptions := data.Options{
```
I thought you are going to use NewOptions here. If not, remove NewOptions?

weinong commented on tests/k6/README.md:
awesome!

weinong commented on server/server.go:
```diff
@@ -228,8 +228,11 @@ func (s Server) ListenAndServe() {
+			cacheOptions := data.NewOptions()
+			cacheOptions.HardMaxCacheSize = s.AuthzRecommendedOptions.Azure.CacheSizeMB
+			cacheOptions.LifeWindow = time.Duration(s.AuthzRecommendedOptions.Azure.CacheTTLMinutes) * time.Minute
+			klog.Infof("Initializing authorization cache: size=%dMB, ttl=%dm", cacheOptions.HardMaxCacheSize, s.AuthzRecommendedOptions.Azure.CacheTTLMinutes)
```
nit(non-blocking): this log will get washed out due to the short retention policy. a common way is to emit metric about the configuration.
